src/trajectory.jl

```diff
@@ -587,79 +525,67 @@ function isterminated(h::Hamiltonian, t::BinaryTree{<:ClassicNoUTurn})
 end
 
 """
-    isterminated(h::Hamiltonian, t::BinaryTree{<:GeneralisedNoUTurn})
+    isterminated(h::Hamiltonian, t::BinaryTree)
 
 Detect U turn for two phase points (`zleft` and `zright`) under given Hamiltonian `h`
 using the generalised no-U-turn criterion.
 
 Ref: https://arxiv.org/abs/1701.02434
 """
-function isterminated(h::Hamiltonian, t::BinaryTree{<:GeneralisedNoUTurn})
-    rho = t.c.rho
+function isterminated(::Union{NoUTurn, StrictNoUTurn}, h::Hamiltonian, t::BinaryTree)
+    rho = t.rho
     s = generalised_uturn_criterion(rho, ∂H∂r(h, t.zleft.r), ∂H∂r(h, t.zright.r))
     return Termination(s, false)
 end
 
+isterminated(tc::Union{ClassicNoUTurn, NoUTurn}, h, t, tleft, tright) = isterminated(tc, h, t)
+
 """
-    isterminated(
-        h::Hamiltonian, t::T, tleft::T, tright::T
-    ) where {T<:BinaryTree{<:StrictGeneralisedNoUTurn}}
+    isterminated(tc::StrictNoUTurn, h, t, tleft, tright)
 
 Detect U turn for two phase points (`zleft` and `zright`) under given Hamiltonian `h`
 using the generalised no-U-turn criterion with additional U-turn checks.
 
-Ref: https://arxiv.org/abs/1701.02434 https://github.com/stan-dev/stan/pull/2800
-"""
-function isterminated(
-    h::Hamiltonian, t::T, tleft::T, tright::T
-) where {T<:BinaryTree{<:StrictGeneralisedNoUTurn}}
-    # Classic generalised U-turn check
-    t_generalised = BinaryTree(
-        t.zleft,
-        t.zright,
-        GeneralisedNoUTurn(t.c.rho),
-        t.sum_α,
-        t.nα,
-        t.ΔH_max
-    )
-    s1 = isterminated(h, t_generalised)
+## References 
 
-    # U-turn checks for left and right subtree
-    # See https://discourse.mc-stan.org/t/nuts-misses-u-turns-runs-in-circles-until-max-treedepth/9727/33
-    # for a visualisation.
+1. https://arxiv.org/abs/1701.02434
```

Comment:
Consider adding the paper title and author information here.

Reply:
Done.